saleor/graphql/discount/dataloaders.py

```diff
@@ -150,6 +151,57 @@ def batch_load(self, keys):
         ]
 
 
+class VoucherInfoByVoucherCodeLoader(DataLoader):
+    context_key = "voucher_info_by_voucher_code"
+
+    def batch_load(self, keys):
+        vouchers_map = (
+            Voucher.objects.using(self.database_connection_name)
+            .filter(code__in=keys)
+            .in_bulk(field_name="code")
+        )
+        vouchers = vouchers_map.values()
+        voucher_products = Voucher.products.through.objects.filter(
+            voucher__in=vouchers
+        ).values_list("voucher_id", "product_id")
+        voucher_variants = Voucher.variants.through.objects.filter(
+            voucher__in=vouchers
+        ).values_list("voucher_id", "productvariant_id")
+        voucher_collections = Voucher.collections.through.objects.filter(
+            voucher__in=vouchers
+        ).values_list("voucher_id", "collection_id")
+        voucher_categories = Voucher.categories.through.objects.filter(
+            voucher__in=vouchers
+        ).values_list("voucher_id", "category_id")
+        product_pks_map = defaultdict(list)
+        variant_pks_map = defaultdict(list)
+        category_pks_map = defaultdict(list)
+        collection_pks_map = defaultdict(list)
+        for voucher_id, product_id in voucher_products:
+            product_pks_map[voucher_id].append(product_id)
+        for voucher_id, variant_id in voucher_variants:
+            variant_pks_map[voucher_id].append(variant_id)
+        for voucher_id, category_id in voucher_categories:
+            category_pks_map[voucher_id].append(category_id)
+        for voucher_id, collection_id in voucher_collections:
+            collection_pks_map[voucher_id].append(collection_id)
+        voucher_infos = []
+        for code in keys:
+            voucher = vouchers_map.get(code)
+            if not voucher:
```

Comment:
This could break the dataloader. Dataloader should always return a list of the same length as `keys`.

Reply:
I used `saleor.graphql.checkout.dataloaders.CheckoutInfoByCheckoutTokenLoader` as an example. `CheckoutInfoByCheckoutTokenLoader` uses `VoucherByCodeLoader` inside the load methods. `VoucherByCodeLoader` fetches a vouchers based on codes taken from checkout:
```
                voucher_codes = {
                    checkout.voucher_code
                    for checkout in checkouts
                    if checkout.voucher_code
                }
                vouchers = VoucherByCodeLoader(self.context).load_many(voucher_codes)
```
In that moment, I don't think that we have an order that could match to `checkouts`. 
After that, we create `voucher_map`:
```
voucher_map = {voucher.code: voucher for voucher in vouchers}
...
voucher = voucher_map.get(checkout.voucher_code)
```
and use it to fetch correct voucher to checkout with promo code. 

I used this flow as an example. @fowczarek What do you think? Do you think this is bad pattern or incorrect usage of dataloader?